### ISSUE TITLE
教师和学生输出层选取原则

### ISSUE THREAD

**Issue by pengsmallcute** [jasonDasuantou/yolov8_distillation#1]
我观察到你的特征比较通道是['6', '8', '12', '15', '18', '21']，这个·似乎是每一层C2F输出的，请问这个选取是你自己随意定的还是有讲究的，

**Comment by garlic-byte** [comment]
因为这几个通道得到的特征比较重要，所以需要学生好好学习

> 我观察到你的特征比较通道是['6', '8', '12', '15', '18', '21']，这个·似乎是每一层C2F输出的，请问这个选取是你自己随意定的还是有讲究的，

这几个特征比较重要，所以让学生好好学习

**Comment by pengsmallcute** [comment]
好的，感谢感谢，但是作者大大，咱们教师模型不是不用更新参数，你咋还再优化器里面加入了教师模型的权重作为可学习的。
        # change new_____________________
        if self.Distillation is not None:
            for v in model1.modules():
                # print(v)
                if hasattr(v, 'bias') and isinstance(v.bias, nn.Parameter):  # bias (no decay)
                    g[2].append(v.bias)
                if isinstance(v, bn):  # weight (no decay)
                    g[1].append(v.weight)
                elif hasattr(v, 'weight') and isinstance(v.weight, nn.Parameter):  # weight (with decay)
                    g[0].append(v.weight)

**Comment by garlic-byte** [comment]
> 好的，感谢感谢，但是作者大大，咱们教师模型不是不用更新参数，你咋还再优化器里面加入了教师模型的权重作为可学习的。
> # change new_____________________
> if self.Distillation is not None:
> for v in model1.modules():
> # print(v)
> if hasattr(v, 'bias') and isinstance(v.bias, nn.Parameter): # bias (no decay)
> g[2].append(v.bias)
> if isinstance(v, bn): # weight (no decay)
> g[1].append(v.weight)
> elif hasattr(v, 'weight') and isinstance(v.weight, nn.Parameter): # weight (with decay)
> g[0].append(v.weight)

我的理解是：教师也需要进行学习才能更好的教育学生！

**Comment by pengsmallcute** [comment]
但是，你在训练阶段使用的是没有梯度回传的。教师模型

&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; with torch.no_grad():
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; pred = self.Distillation(batch['img']) 





------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年8月11日(星期天) 中午1:10
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [jasonDasuantou/yolov8_distillation] 教师和学生输出层选取原则 (Issue #1)














  
好的，感谢感谢，但是作者大大，咱们教师模型不是不用更新参数，你咋还再优化器里面加入了教师模型的权重作为可学习的。
 
change new_____________________
 
if self.Distillation is not None:
 for v in model1.modules():
 
print(v)
 
if hasattr(v, 'bias') and isinstance(v.bias, nn.Parameter): # bias (no decay)
 g[2].append(v.bias)
 if isinstance(v, bn): # weight (no decay)
 g[1].append(v.weight)
 elif hasattr(v, 'weight') and isinstance(v.weight, nn.Parameter): # weight (with decay)
 g[0].append(v.weight)
  
我的理解是：教师也需要进行学习才能更好的教育学生！

**Comment by garlic-byte** [comment]
> 但是，你在训练阶段使用的是没有梯度回传的。教师模型 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; with torch.no_grad(): &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; pred = self.Distillation(batch['img'])
> […](#)

稍等，我得去做个调研。如果你知道问题答案也请告诉我。

**Comment by pengsmallcute** [comment]
我觉得比较合适的情况是，教师模型冻结，
我们训练的教师模型应该是已经完全收敛的，
这个地方不应该再更新教师模型参数，
而是专注于训练学生模型。

并且，我有一点疑惑的是在使用MGD_LOSS的时候，调整通道的self.generation 中带有可学习参数的卷积，这会不会影响学习结果，还是应该固定卷积参数。
class MGDLoss(nn.Module):
&nbsp; &nbsp; def __init__(self, channels_s, channels_t, alpha_mgd=0.00002, lambda_mgd=0.65):
&nbsp; &nbsp; &nbsp; &nbsp; super(MGDLoss, self).__init__()
&nbsp; &nbsp; &nbsp; &nbsp; device = 'cuda' if torch.cuda.is_available() else 'cpu'

&nbsp; &nbsp; &nbsp; &nbsp; self.alpha_mgd = alpha_mgd
&nbsp; &nbsp; &nbsp; &nbsp; self.lambda_mgd = lambda_mgd

&nbsp; &nbsp; &nbsp; &nbsp; self.generation = [
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; nn.Sequential(
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; nn.Conv2d(channel_s, channel, kernel_size=3, padding=1),
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; nn.ReLU(inplace=True),
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; nn.Conv2d(channel, channel, kernel_size=3, padding=1)).to(device) for channel_s,channel in zip(channels_s,channels_t)
&nbsp; &nbsp; &nbsp; &nbsp; ] 







------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年8月11日(星期天) 中午1:23
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [jasonDasuantou/yolov8_distillation] 教师和学生输出层选取原则 (Issue #1)














  
但是，你在训练阶段使用的是没有梯度回传的。教师模型 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; with torch.no_grad(): &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; pred = self.Distillation(batch['img'])
 …
  
稍等，我得去做个调研。如果你知道问题答案也请告诉我。

**Comment by garlic-byte** [comment]
> 稍等，我得去做个调研。如果你知道问题答案也请告诉我。

我在后面代码中，教师和学生的loss应该是有回传的吧。

**Comment by pengsmallcute** [comment]
但是，你没有在训练阶段推理部分保证教师模型梯度，所以梯度无法回传的·

**Comment by garlic-byte** [comment]
> 无法回传的·

能否告诉我，哪一种是保证模型梯度，哪一种是没有保证？最好有实际的例子，谢谢！

**Comment by pengsmallcute** [comment]
你使用了torch.no_grad命令，这个不会计算教师模型梯度
![PixPin_2024-08-13_13-15-45](https://github.com/user-attachments/assets/6852ff21-20e6-4eec-a6d4-bbacd3009dfc)

**Comment by peace-1314** [comment]
剪枝后的学生模型的18层通道数不一致，导致报错，这个怎么解决
Layer 140: model.18.cv1.conv has 128 output channels
Layer 143: model.18.cv2.conv has 104 output channels

**Comment by garlic-byte** [comment]
你试着调整一下，是不是学生模型的通道数不一样。
&nbsp; &nbsp; &nbsp; &nbsp; channels_s = [256, 480, 256, 64, 143, 229][-le:]
&nbsp; &nbsp; &nbsp; &nbsp; channels_t = [256, 512, 256, 128, 256, 512][-le:]

如果还不行，你可以看一下我的博客：yolov8蒸馏（附代码-免费）_yolov8剪枝和蒸馏-CSDN博客
或者把你的报错代码全部发给我看一下。


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "jasonDasuantou/yolov8_distillation"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年8月29日(星期四) 晚上10:15
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [jasonDasuantou/yolov8_distillation] 教师和学生输出层选取原则 (Issue #1)





 
剪枝后的学生模型的18层通道数不一致，导致报错，这个怎么解决
 Layer 140: model.18.cv1.conv has 128 output channels
 Layer 143: model.18.cv2.conv has 104 output channels
